File: sdk/samples/WFSamples/Applications/WorkflowMonitor/VB/WorkflowMonitor/DatabaseService.vb
'---------------------------------------------------------------------
'  This file is part of the Windows Workflow Foundation SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'---------------------------------------------------------------------

Imports System
Imports System.Globalization
Imports System.Workflow.Runtime.Tracking


' This class provides database query services against the TrackingService
' via several stored procedure that must be installed prior to running
' the monitor.
Public Class DatabaseService
    Private connectionString As String
    Private serverNameValue As String = String.Empty
    Private databaseNameValue As String = String.Empty

    Friend Sub New()
        connectionString = "Persist Security Info=False;Integrated Security=SSPI;database=" + databaseNameValue + ";server=" + serverNameValue
    End Sub

    Friend Property ServerName() As String
        Get
            Return serverNameValue
        End Get
        Set(ByVal Value As String)
            serverNameValue = Value
            connectionString = "Persist Security Info=False;Integrated Security=SSPI;database=" + databaseNameValue + ";server=" + serverNameValue
        End Set
    End Property

    Friend Property DatabaseName() As String
        Get
            Return databaseNameValue
        End Get
        Set(ByVal Value As String)
            databaseNameValue = Value
            connectionString = "Persist Security Info=False;Integrated Security=SSPI;database=" + databaseNameValue + ";server=" + serverNameValue

        End Set
    End Property


    ' Query for all workflows from the SqlTrackingService using the 'GetWorkflowStatus' stored procedure
    Friend Function GetWorkflows( _
            ByVal workflowEvent As String, _
            ByVal from As System.DateTime, _
            ByVal until As DateTime, _
            ByVal trackingDataItemValue As TrackingDataItemValue) _
            As List(Of SqlTrackingWorkflowInstance)
        Try
            Dim queriedWorkflows As List(Of SqlTrackingWorkflowInstance) = New List(Of SqlTrackingWorkflowInstance)()
            Dim sqlTrackingQuery As SqlTrackingQuery = New SqlTrackingQuery(connectionString)
            Dim sqlTrackingQueryOptions As SqlTrackingQueryOptions = New SqlTrackingQueryOptions()
            sqlTrackingQueryOptions.StatusMinDateTime = from.ToUniversalTime()
            sqlTrackingQueryOptions.StatusMaxDateTime = until.ToUniversalTime()
            ' If QualifiedName, FieldName, or DataValue is not supplied, we will not query since they are all required to match
            If (Not ((trackingDataItemValue.QualifiedName = String.Empty) Or (trackingDataItemValue.FieldName = String.Empty) Or ((trackingDataItemValue.DataValue = String.Empty)))) Then
                sqlTrackingQueryOptions.TrackingDataItems.Add(trackingDataItemValue)
            End If

            queriedWorkflows.Clear()
            If (workflowEvent.ToLower(CultureInfo.InvariantCulture) = "created") Then
                sqlTrackingQueryOptions.WorkflowStatus = System.Workflow.Runtime.WorkflowStatus.Created
                queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions))
            ElseIf (workflowEvent.ToLower(CultureInfo.InvariantCulture) = "completed") Then
                sqlTrackingQueryOptions.WorkflowStatus = System.Workflow.Runtime.WorkflowStatus.Completed
                queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions))
            ElseIf (workflowEvent.ToLower(CultureInfo.InvariantCulture) = "running") Then
                sqlTrackingQueryOptions.WorkflowStatus = System.Workflow.Runtime.WorkflowStatus.Running
                queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions))
            ElseIf (workflowEvent.ToLower(CultureInfo.InvariantCulture) = "suspended") Then
                sqlTrackingQueryOptions.WorkflowStatus = System.Workflow.Runtime.WorkflowStatus.Suspended
                queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions))
            ElseIf (workflowEvent.ToLower(CultureInfo.InvariantCulture) = "terminated") Then
                sqlTrackingQueryOptions.WorkflowStatus = System.Workflow.Runtime.WorkflowStatus.Terminated
                queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions))
            ElseIf ((workflowEvent = Nothing) Or _
                    (workflowEvent.ToLower(CultureInfo.InvariantCulture) = "all") Or _
                    (workflowEvent = String.Empty)) Then
                sqlTrackingQueryOptions.WorkflowStatus = System.Workflow.Runtime.WorkflowStatus.Created
                queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions))

                sqlTrackingQueryOptions.WorkflowStatus = System.Workflow.Runtime.WorkflowStatus.Completed
                queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions))

                sqlTrackingQueryOptions.WorkflowStatus = System.Workflow.Runtime.WorkflowStatus.Running
                queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions))

                sqlTrackingQueryOptions.WorkflowStatus = System.Workflow.Runtime.WorkflowStatus.Suspended
                queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions))

                sqlTrackingQueryOptions.WorkflowStatus = System.Workflow.Runtime.WorkflowStatus.Terminated
                queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions))

            End If

            Return queriedWorkflows

        Catch exception As Exception
            'Dim errorMessage As String = "Exception in GetWorkflows" + vbCrLf + "Database: " + databaseName + vbCrLf + "Server: " + serverName
            Throw (New Exception("Exception in GetWorkflows", exception))
        End Try
    End Function

    Friend Function GetWorkflow(ByVal workflowInstanceId As Guid) As SqlTrackingWorkflowInstance

        Try
            Dim sqlTrackingQuery As SqlTrackingQuery = New SqlTrackingQuery(connectionString)
            Dim sqlTrackingWorkflowInstance As SqlTrackingWorkflowInstance = Nothing
            sqlTrackingQuery.TryGetWorkflow(workflowInstanceId, sqlTrackingWorkflowInstance)
            Return sqlTrackingWorkflowInstance
        Catch ex As Exception
            Throw New Exception("Exception in GetWorkflow", ex)
        End Try
    End Function
End Class

File: sdk/samples/WFSamples/Applications/WorkflowMonitor/CS/WorkflowMonitor/DatabaseService.cs
//---------------------------------------------------------------------
//  This file is part of the Windows Workflow Foundation SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------
using System;
using System.Collections.Generic;
using System.Workflow.Runtime.Tracking;
using System.Workflow.Runtime;
using System.Globalization;

namespace Microsoft.Samples.Workflow.WorkflowMonitor
{
    //This class provides database query services against the TrackingService using SqlTrackingQuery APIs
    internal sealed class DatabaseService
    {
        private string connectionString;
        private string serverNameValue = String.Empty;
        private string databaseNameValue = String.Empty;

        internal DatabaseService()
        {
            connectionString = "Persist Security Info=False;Integrated Security=SSPI;database=" + databaseNameValue + ";server=" + serverNameValue;
        }

        internal string ServerName
        {
            get
            {
                return serverNameValue;
            }
            set
            {
                serverNameValue = value;
                connectionString = "Persist Security Info=False;Integrated Security=SSPI;database=" + databaseNameValue + ";server=" + serverNameValue;
            }
        }

        internal string DatabaseName
        {
            get
            {
                return databaseNameValue;
            }
            set
            {
                databaseNameValue = value;
                connectionString = "Persist Security Info=False;Integrated Security=SSPI;database=" + databaseNameValue + ";server=" + serverNameValue;
            }
        }

        internal List<SqlTrackingWorkflowInstance> GetWorkflows(string workflowEvent, DateTime from, DateTime until, TrackingDataItemValue trackingDataItemValue)
        {
            try
            {
                List<SqlTrackingWorkflowInstance> queriedWorkflows = new List<SqlTrackingWorkflowInstance>();
                SqlTrackingQuery sqlTrackingQuery = new SqlTrackingQuery(connectionString);
                SqlTrackingQueryOptions sqlTrackingQueryOptions = new SqlTrackingQueryOptions();
                sqlTrackingQueryOptions.StatusMinDateTime = from.ToUniversalTime();
                sqlTrackingQueryOptions.StatusMaxDateTime = until.ToUniversalTime();
                // If QualifiedName, FieldName, or DataValue is not supplied, we will not query since they are all required to match
                if (!((string.Empty == trackingDataItemValue.QualifiedName) || (string.Empty == trackingDataItemValue.FieldName) || ((string.Empty == trackingDataItemValue.DataValue))))
                    sqlTrackingQueryOptions.TrackingDataItems.Add(trackingDataItemValue);

                queriedWorkflows.Clear();

                if ("created" == workflowEvent.ToLower(CultureInfo.InvariantCulture))
                {
                    sqlTrackingQueryOptions.WorkflowStatus = WorkflowStatus.Created;
                    queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions));
                }
                else if ("completed" == workflowEvent.ToLower(CultureInfo.InvariantCulture))
                {
                    sqlTrackingQueryOptions.WorkflowStatus = WorkflowStatus.Completed;
                    queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions));
                }
                else if ("running" == workflowEvent.ToLower(CultureInfo.InvariantCulture))
                {
                    sqlTrackingQueryOptions.WorkflowStatus = WorkflowStatus.Running;                    
                    queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions));
                }
                else if ("suspended" == workflowEvent.ToLower(CultureInfo.InvariantCulture))
                {
                    sqlTrackingQueryOptions.WorkflowStatus = WorkflowStatus.Suspended;
                    queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions));
                }
                else if ("terminated" == workflowEvent.ToLower(CultureInfo.InvariantCulture))
                {
                    sqlTrackingQueryOptions.WorkflowStatus = WorkflowStatus.Terminated;
                    queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions));
                }
                else if (("all" == workflowEvent.ToLower(CultureInfo.InvariantCulture)) || (string.Empty == workflowEvent) || (null == workflowEvent))
                {
                    sqlTrackingQueryOptions.WorkflowStatus = WorkflowStatus.Created;
                    queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions));

                    sqlTrackingQueryOptions.WorkflowStatus = WorkflowStatus.Completed;
                    queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions));

                    sqlTrackingQueryOptions.WorkflowStatus = WorkflowStatus.Running;
                    queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions));

                    sqlTrackingQueryOptions.WorkflowStatus = WorkflowStatus.Suspended;
                    queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions));

                    sqlTrackingQueryOptions.WorkflowStatus = WorkflowStatus.Terminated;
                    queriedWorkflows.AddRange(sqlTrackingQuery.GetWorkflows(sqlTrackingQueryOptions));

                }
                return queriedWorkflows;

            }
            catch (Exception exception)
            {
                throw new Exception("Exception in GetWorkflows", exception);
            }
        }

        internal bool TryGetWorkflow(Guid workflowInstanceId, out SqlTrackingWorkflowInstance sqlTrackingWorkflowInstance)
        {
            SqlTrackingQuery sqlTrackingQuery = new SqlTrackingQuery(connectionString);
            return sqlTrackingQuery.TryGetWorkflow(workflowInstanceId, out sqlTrackingWorkflowInstance);
        }
    }
}
